daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/ChoiceCombinator.scala

```diff
@@ -0,0 +1,124 @@
+ *
+ * For direct dispatch, we create a disapatch-branch key map
+ * which is used to determine which branch to parse at runtime.
+ *
+ * In the unparser case, we know which element we got from the infoset, but we
+ * need to determine which branch of the choice to take at runtime. This
+ * unparser uses a Map to make the determination based on the element seen.
+ */
+case class ChoiceCombinator(ch: ChoiceTermBase, rawAlternatives: Seq[Gram])
+  extends Terminal(ch, !rawAlternatives.isEmpty) {
+
+  private lazy val alternatives = rawAlternatives.filterNot(_.isEmpty)
+
+  private lazy val parsers = alternatives.map { _.parser }.filterNot { _.isEmpty }
+
+  override def isEmpty = super.isEmpty || alternatives.isEmpty
+
+  lazy val parser: Parser = {
+    if (!ch.isDirectDispatch) {
+      new ChoiceParser(ch.termRuntimeData, parsers)
+    } else {
+      val dispatchBranchKeyValueTuples = alternatives.flatMap { alt =>
+        val keyTerm = alt.context.asInstanceOf[Term]
+        val cbk = keyTerm.choiceBranchKey
+        // Note that this behaves differently than the specification, since
+        // this accepts a space separated list of keys
+        val cbks = ChoiceBranchKeyCooker.convertConstant(cbk, keyTerm.runtimeData, forUnparse = false)
+        cbks.map { (_, alt) }
+      }
+
+      // check for duplicate dfdl:choiceBranchKeys
+      val groupedByKey = dispatchBranchKeyValueTuples.groupBy(_._1)
+      groupedByKey.foreach {
+        case (k, kvs) =>
+          if (kvs.length > 1) {
+            SDE("dfdl:choiceBranchKey value (%s) is not unique across all branches of a direct dispatch choice. Offending branches are:\n%s",
+              k, kvs.map(_._2.context.runtimeData).map(rd => rd.diagnosticDebugName + " " + rd.schemaFileLocation.locationDescription).mkString("- ", "\n- ", ""))
+          }
+      }
+
+      val dispatchBranchKeyMap = dispatchBranchKeyValueTuples.toMap.mapValues(_.parser)
+      val serializableMap = dispatchBranchKeyMap.map(identity)
+
+      new ChoiceDispatchCombinatorParser(ch.termRuntimeData, ch.choiceDispatchKeyEv, serializableMap)
+    }
+  }
+
+  override lazy val unparser: Unparser = {
+    if (!ch.isHidden) {
+      val eventRDMap = ch.choiceBranchMap
+      val eventUnparserMap = eventRDMap.flatMap {
+        case (cbe, rd) =>
+          // if we don't find a matching RD for a term that's probably
+          // because the term is an empty sequence or empty choice (which do happen
+          // and we even have tests for them). Since those can never be chosen by
+          // means of an element event, they don't appear in the map.
+          val altGram = alternatives.find { alt =>
+            val crd = alt.context.runtimeData
+            val found = crd =:= rd
+            found
+          }
+          altGram.map { ag => (cbe, ag.unparser) }
+      }
+      val mapValues = eventUnparserMap.map { case (k, v) => v }.toSeq.filterNot(_.isEmpty)
+      if (mapValues.isEmpty)
```

Comment:
So this supports an invariant behavior that Grammar terms now must support which is that they optimize when producing parser and unparser by looking to see if they can optimize away to NadaParser or NadaUnparser.  Nothing is allowed to just punt and execute calling NadaParser or NadaUnparser or passing those to an underlying combinator parser/unparser.  I verified that NadaParser now faults when run. Looks like NadaUnparser still is a "no-op" at runtime. But I think that's an oversight. (I will try changing it and see if anything fails.) 

So like I say, it's a behavior invariant for grammar terms now that is intended to avoid the need to special case lots of corner situations, but more .... really.... to catch places where the generation of parser/unparser isn't adequately thought through and NadaParser or NadaUnparser are just being left in as arguments to some parse/unparse combinator or other. All such combinators now if they have an optional parser that may or may not be there, they have a sequence/vector type or a Maybe type to explicitly deal with it. They don't just assume they get a parser, and invoke it, depending on users of that combinator to put in NadaParser if they don't want the combinator to do anything.